fabric/src/test/java/TestStyledText.java

```diff
@@ -0,0 +1,413 @@
+        }
+    }
+
+    @Test
+    public void normalizedStringCreation_shouldProduceCorrectString() {
+        final String badText = "ÀÀÀÀHello, ÀWorld!֎";
+        final String expected = "Hello, World!";
+
+        StyledText styledText = StyledText.fromString(badText);
+
+        Assertions.assertEquals(
+                expected,
+                styledText.getNormalized().getString(PartStyle.StyleType.NONE),
+                "StyledText.getNormalized().getString() returned an unexpected value.");
+    }
+
+    @Test
+    public void untrimmedText_shouldProduceTrimmedString() {
+        final String badText = "   Hello, World!  ";
+        final String expected = "Hello, World!";
+
+        StyledText styledText = StyledText.fromString(badText);
+
+        Assertions.assertEquals(
+                expected,
+                styledText.trim().getString(PartStyle.StyleType.NONE),
+                "StyledText.getString() returned an unexpected value.");
+    }
+
+    @Test
+    public void emptyText_shouldProduceEmptyString() {
+        StyledText styledText = StyledText.fromString("");
+
+        Assertions.assertTrue(styledText.isEmpty(), "StyledText.isEmpty() returned an unexpected value.");
+    }
+
+    @Test
+    public void blankText_shouldProduceBlankString() {
+        StyledText styledText = StyledText.fromString("    ");
+
+        Assertions.assertTrue(styledText.isBlank(), "StyledText.isBlank() returned an unexpected value.");
+    }
+
+    @Test
+    public void contains_shouldProduceCorrectResult() {
+        final String testString = "§c§oitalicred§9§oblue§cnonitalic§oinherited§lbold§rafter";
+        final String[] expectedParts = {"§c§oitalicred", "§9§oblue", "§cnonitalic", "§oinherited", "§lbold", "§rafter"};
+
+        StyledText styledText = StyledText.fromString(testString);
+
+        for (String expectedPart : expectedParts) {
+            Assertions.assertTrue(
+                    styledText.contains(expectedPart),
+                    "StyledText.contains() could not find a part matching: " + expectedPart);
+        }
+
+        for (String expectedPart : expectedParts) {
+            Assertions.assertTrue(
+                    styledText.contains(StyledText.fromString(expectedPart)),
+                    "StyledText.contains() could not find a part matching: " + expectedPart);
+            Assertions.assertTrue(
+                    styledText.contains(StyledText.fromString(expectedPart), PartStyle.StyleType.NONE),
+                    "StyledText.contains(NONE) could not find a part matching: " + expectedPart);
+        }
+    }
+
+    @Test
+    public void startsWith_shouldProduceCorrectResult() {
+        final String testString = "§c§oitalicred§r§9§oblue§r§cnonitalic§oinherited§lbold§rafter";
+        final String expectedStart = "§c§oitalicred";
+
+        StyledText styledText = StyledText.fromString(testString);
```

Comment:
Same here and for endsWith, we want to make sure it also works across part breaks, so like "after" and "old§rafter" as well. And perhaps, what should "" return? true or false? Not terribly important, but looking for edge cases is a great way of honing your test writing skills...

Reply:
I can add the special cases. About "", it will return what the string impl does. So we get the result we would like to "follow".